PathFilterUI : Drop onto Box and Reference nodes

ericmehl opened the pull request:
This adds the ability to drag and drop scene paths onto Box and Reference nodes. I can think of a few complications, like having multiple filters connected to a Box / Reference, that might not make this ideal. I believe the request came from the lighting teams and my recent work on #5786 had me in the frame of mind to bubble up the request on my list.

If it proves to be more problematic than beneficial, I'm fine with dropping it, but might be worth pinging the lighting team first to see how often they think it would help.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

murraystevenson commented:
One limitation of the simple approach is that we now expect every Box and Reference to have a "filter" plug. 

Dragging paths over a Box/Reference without one results in the below. 

```
ERROR :   File "gaffer-pr5796-dragDropReferenceBox-2024_04_15_2009-a86e04c9-linux-gcc11/python/GafferSceneUI/PathFilterUI.py", line 307, in __dragEnter
ERROR :     if __dropMode( nodeGadget, event ) == __DropMode.None_ :
ERROR :   File "gaffer-pr5796-dragDropReferenceBox-2024_04_15_2009-a86e04c9-linux-gcc11/python/GafferSceneUI/PathFilterUI.py", line 255, in __dropMode
ERROR :     if nodeGadget.node()["filter"].getInput() is not None :
ERROR : KeyError: "'filter' is not a child of 'Box'"
```

Dropping paths onto a Box with a plug named "filter" that isn't a `FilterPlug` will create a PathFilter and then error while attempting to connect it to the incompatible plug type.

murraystevenson commented:
In solving this, we'd also want to remove this listed limitation from the documentation: https://github.com/GafferHQ/gaffer/blob/b5e245463baeec15878e777201b4ac76a8bb54f4/doc/source/WorkingWithTheNodeGraph/BoxNode/index.md?plain=1#L264-L266

ericmehl commented:
Thanks for pointing out the documentation change that needs to be made and testing the functionality @murraystevenson. I think I have a pretty good working version with the addition of https://github.com/GafferHQ/gaffer/pull/5796/commits/df7fbb23f3e5e7a61e540d06ec80e0fb776eb84f.

Now, for any node that has been registered for the drag and drop events, it will look at all the child filter nodes and allow you to drop onto it if there is exactly one. I think that's a pretty reasonable constraint for a convenience feature like this?

ericmehl commented:
As we talked about live this afternoon, I changed the node type the drag handling targets to `Gaffer.SubGraph` instead of the pair of `Gaffer.Box` and `Gaffer.Reference`. Otherwise it's all the same so I went ahead and squashed down.

The behavior seems to be working as expected still, ready for a new look.

murraystevenson reviewed (approved):
Thanks Eric! The updated behaviour looks good to me.
